libs/community/langchain_community/vectorstores/cloudSQL.py

```diff
@@ -0,0 +1,330 @@
+from __future__ import annotations
+
+import uuid
+
+from typing import Any, Callable, Dict, List, Optional, Tuple, Type
+
+from langchain_core.documents import Document
+from langchain_core.vectorstores import VectorStore
+from langchain_core.embeddings import Embeddings
+
+from google.cloud.sql.connector import Connector
+import google.auth
+
+import sqlalchemy
+from sqlalchemy import inspect
+from sqlalchemy import Table, Column, Integer, String
+from sqlalchemy.dialects.postgresql import JSON, UUID
+
+from pgvector.sqlalchemy import Vector
+
+def _get_IAM_user(credentials):
+"""Get user/service account name"""
+import google.auth
+from google.auth.transport.requests import Request
+from google.auth.compute_engine import _metadata
+
+# Checks for service account
+if hasattr(credentials, "service_account_email"):
+    return credentials.service_account_email[:-20]
+else:
+    # Defaults to getting the user account
+    command = ["gcloud", "auth", "list", "--filter=status:ACTIVE", "--format=value(account)"]
+    result = subprocess.run(command, stdout=subprocess.PIPE, text=True)
+    if result.returncode == 0:
+        return result.stdout.strip()
+    else:
+        return f"Error: {result.stderr}"
+        
+class CloudSQLEngine:
+    """Creating a connection to the CloudSQL instance
+    To use, you need the following packages installed:
+        cloud-sql-python-connector[asyncpg]
+    """
+    
+    connector = Connector()
+    
+    @staticmethod
+    def from_instance(
+        region: str, 
+        instance: str, 
+        database: str, 
+        project_id: str=None
+    ) -> sqlalchemy.engine.Engine:
+     """Create sqlalchemy connection to the postgres database in the CloudSQL instance.
+
+        Args:
+            region (str): CloudSQL instance region.
+            instance (str): CloudSQL instance name.
+            database (str): CloudSQL instance database name.
+            project_id (str): GCP project ID. Defaults to None
+
+        Returns:
+            Sqlalchemy engine containg the connection pool.
+        """
+        
+        if project_id is None:
+            credentials, project_id = google.auth.default()
+
+        IAM_USER = _get_IAM_user(credentials)
+        
+        conn = connector.connect(
+            INSTANCE_URI = f"{project_id}:{region}:{instance}",
+            'asyncpg',
+            user = IAM_USER,
+            db = database,
+            enable_iam_auth = True
+        )
+        
+        pool = sqlalchemy.create_engine("postgresql+asyncpg://",creator=conn)
```

Comment:
We need to use create_async_engine with async_creator. async_creator takes in a function, see example https://github.com/GoogleCloudPlatform/genai-databases-retrieval-app/blob/f03970b23a79ae66f848f6a83558530820ecab8a/retrieval_service/datastore/providers/cloudsql_postgres.py#L58